Add a placeholder option to the Select input

samyou-softwire opened the pull request:
adds the option for a default option to be added to the select box

this will have the value `select_placeholder_value` and the display text is provided by the user

so that the user does not need to program a handler for the placeholder value a binder is provided as well that will interpret the default value as null

samyou-softwire committed:
Add a placeholder to the select input

uses a fixed value for the placeholder

samyou-softwire committed:
Add a model binder for the select input

this will null the placeholder value if selected, which will trigger required checks on the caller

jdgage commented:
This will need testing with a screen-reader/other assistive technology. What may happen is it reads out `select_placeholder_value`, which will be strange to users. Looking at the GDS [example](https://design-system.service.gov.uk/components/select/), they use a value that reflects the displayed text. I think we should allow consumers to control the value, so that they can do something like that.

samyou-softwire commented:
> This will need testing with a screen-reader/other assistive technology. What may happen is it reads out `select_placeholder_value`, which will be strange to users. Looking at the GDS [example](https://design-system.service.gov.uk/components/select/), they use a value that reflects the displayed text. I think we should allow consumers to control the value, so that they can do something like that.

I did a quick check on NVDA & Screen reader on Android and the labels were read out by both & not the value. I'm not sure why a reader would read out this info that a sighted user would not see?

we could let a user specify their own placeholder but this seems like more work for the caller to have to register their own placeholder string and plug this into the model binder. I'm modelling this mostly off of the checkboxes which have `hidden_checkbox_dummy_value` set up to work with their model binder

jdgage commented:
> > This will need testing with a screen-reader/other assistive technology. What may happen is it reads out `select_placeholder_value`, which will be strange to users. Looking at the GDS [example](https://design-system.service.gov.uk/components/select/), they use a value that reflects the displayed text. I think we should allow consumers to control the value, so that they can do something like that.
> 
> I did a quick check on NVDA & Screen reader on Android and the labels were read out by both & not the value. I'm not sure why a reader would read out this info that a sighted user would not see?
> 
> we could let a user specify their own placeholder but this seems like more work for the caller to have to register their own placeholder string and plug this into the model binder. I'm modelling this mostly off of the checkboxes which have `hidden_checkbox_dummy_value` set up to work with their model binder

See [here](https://a11ysupport.io/tech/html/option_element)

samyou-softwire commented:
> See [here](https://a11ysupport.io/tech/html/option_element)

looked into it this, though the only reference I could see is that the value may be read out in a datalist context. I think this refers to the html `<datalist>` element, which we don't use here, so I think it should be fine?

jdgage commented:
> > See [here](https://a11ysupport.io/tech/html/option_element)
> 
> looked into it this, though the only reference I could see is that the value may be read out in a datalist context. I think this refers to the html `<datalist>` element, which we don't use here, so I think it should be fine?

That link is a spec for assistive tools: there's a line in there under `Expectation: convey its name` that suggests there are screen-readers that will read the value. These tools are often highly configurable and users will user them in their own personal ways - it's a trap trying to second-guess what they'll do, which is why it's important to write unsurprising, semantically meaningful HTML, so that assistive tools will do what their users expect them to do.

samyou-softwire commented:
> That link is a spec for assistive tools: there's a line in there under `Expectation: convey its name` that suggests there are screen-readers that will read the value. These tools are often highly configurable and users will user them in their own personal ways - it's a trap trying to second-guess what they'll do, which is why it's important to write unsurprising, semantically meaningful HTML, so that assistive tools will do what their users expect them to do.

ahh yeah now I see what you're referencing. I'll let the user specify

for the binder, I could look into making it take in a placeholder, though at this point the binder is not specific to a select input at all (the flow is if = some value, replace with null), so will probably let this be a more generically named binder which could be used for other picker components

samyou-softwire committed:
Allow for the select to take a user specified placeholder value

made the code more generic, could be used for any placeholders the user should not submit

samyou-softwire committed:
Require the user to specify their own placeholder value

this allows for a meaningful value to be entered as this may be read out by screen readers

samyou-softwire commented:
@jdgage I've made updates now to allow for specifying your own placeholder value, modelled mostly on [this](https://stackoverflow.com/a/59921799/23134352). does add a little extra machinery but not too bad. [PR in our repo](https://github.com/UKGovernmentBEIS/desnz-warm-homes-local-grant/pull/157) contains an example usage

jdgage reviewed (commented):
Some things to think about:
1. Can the custom value be made optional (falling back to `placeholder` perhaps?)? Important that we let consumers customise it, but they should be able to opt out of the machinery + boilerplate if they want. Maybe this is how it is already - I can't quite tell?
2. How is all this best documented, library-side?

samyou-softwire committed:
Allow for a default placeholder if a custom one is not specified

document usage of the binder a bit more thoroughly

samyou-softwire commented:
> Can the custom value be made optional (falling back to placeholder perhaps?)? Important that we let consumers customise it, but they should be able to opt out of the machinery + boilerplate if they want. Maybe this is how it is already - I can't quite tell?

I've added this function, I've let it fall back to the GDS default "choose" if one is not specified

> How is all this best documented, library-side?

I've added some more thorough documentation to the binder, the least intuitive part

samyou-softwire committed:
Require only one attribute to be applied

have the placeholder attribute inherit the binder to reduce boilerplate for the user

change its name and documentation to reflect this

samyou-softwire committed:
Ensure documentation is included in the package

samyou-softwire commented:
fwiw it looks like here the documentation isn't being exported on package. have now turned that on

jdgage commented:
Nice!
